idx/memory/partitioned_idx.go

```diff
@@ -0,0 +1,536 @@
+package memory
+
+import (
+	"context"
+	"sort"
+	"time"
+
+	"github.com/grafana/metrictank/cluster"
+	"github.com/grafana/metrictank/idx"
+	"github.com/raintank/schema"
+	log "github.com/sirupsen/logrus"
+	"golang.org/x/sync/errgroup"
+)
+
+// Implements the the "MetricIndex" interface
+type PartitionedMemoryIdx struct {
+	Partition map[int32]*MemoryIdx
+}
+
+func NewPartitionedMemoryIdx() *PartitionedMemoryIdx {
+	idx := &PartitionedMemoryIdx{
+		Partition: make(map[int32]*MemoryIdx),
+	}
+	partitions := cluster.Manager.GetPartitions()
+	//log.Infof("PartitionedMemoryIdx: initializing with partitions: %v", partitions)
+	for _, p := range partitions {
+		idx.Partition[p] = NewMemoryIdx()
+	}
+	return idx
+}
+
+// Init initializes the index at startup and
+// blocks until the index is ready for use.
+func (p *PartitionedMemoryIdx) Init() error {
+	for _, m := range p.Partition {
+		//log.Infof("PartitionedMemoryIdx: initializing MemoryIdx for partition %d", part)
+		err := m.Init()
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+// Stop shuts down the index.
+func (p *PartitionedMemoryIdx) Stop() {
+	for _, m := range p.Partition {
+		m.Stop()
+	}
+}
+
+// Update updates an existing archive, if found.
+// It returns whether it was found, and - if so - the (updated) existing archive and its old partition
+func (p *PartitionedMemoryIdx) Update(point schema.MetricPoint, partition int32) (idx.Archive, int32, bool) {
+	return p.Partition[partition].Update(point, partition)
+}
+
+// AddOrUpdate makes sure a metric is known in the index,
+// and should be called for every received metric.
+func (p *PartitionedMemoryIdx) AddOrUpdate(mkey schema.MKey, data *schema.MetricData, partition int32) (idx.Archive, int32, bool) {
+	return p.Partition[partition].AddOrUpdate(mkey, data, partition)
+}
+
+// UpdateArchive updates the archive information
+func (p *PartitionedMemoryIdx) UpdateArchive(archive idx.Archive) {
+	p.Partition[archive.Partition].UpdateArchive(archive)
+}
+
+// Get returns the archive for the requested id.
+func (p *PartitionedMemoryIdx) Get(key schema.MKey) (idx.Archive, bool) {
+	g, _ := errgroup.WithContext(context.Background())
+	resultChan := make(chan *idx.Archive, len(p.Partition))
+	for _, m := range p.Partition {
+		m := m
+		g.Go(func() error {
+			if a, ok := m.Get(key); ok {
+				resultChan <- &a
+			}
+			return nil
+		})
+	}
+	if err := g.Wait(); err != nil {
+		log.Errorf("memory-idx: failed to get Archive with key %v. %s", key, err)
+		return idx.Archive{}, false
+	}
+	close(resultChan)
+	result, ok := <-resultChan
+	if !ok {
+		return idx.Archive{}, false
+	}
+
+	return *result, true
+}
+
+// GetPath returns the archives under the given path.
+func (p *PartitionedMemoryIdx) GetPath(orgId uint32, path string) []idx.Archive {
+	g, _ := errgroup.WithContext(context.Background())
+	result := make([][]idx.Archive, len(p.Partition))
```

Comment:
is there any scenario where a key would be in more than one index partition? if not, then a length of 1 should be sufficient. if the reason for len(p.Partition) is to not block the insertion into the chan:

further down i think it would make more sense to only wait for the first result in the chan and then not wait for the other routines to finish, otherwise we might end up waiting for a search to complete even though we already have the result. Unfortunately I can't see how to do that with `errgroup.WithContext`.